crates/sail-common/src/spec/data_type.rs

```diff
@@ -1,74 +1,219 @@
+    /// Corresponds to [arrow_schema::DataType::FixedSizeList].
+    FixedSizeList {
+        field: FieldRef,
+        length: i32,
     },
+    /// A list of some logical data type with variable length and 64-bit offsets.
+    /// A single LargeList array can store up to [`i64::MAX`] elements in total.
+    /// Corresponds to [arrow_schema::DataType::LargeList].
+    LargeList {
+        field: FieldRef,
+    },
+    /// A nested datatype that contains a number of sub-fields.
+    /// Corresponds to [arrow_schema::DataType::Struct].
     Struct {
         fields: Fields,
     },
+    /// A nested datatype that can represent slots of differing types.
+    /// Corresponds to [arrow_schema::DataType::Union].
+    Union {
+        union_fields: UnionFields,
+        union_mode: UnionMode,
+    },
+    /// A dictionary encoded array (`key_type`, `value_type`), where each array element is an index
+    /// of `key_type` into an associated dictionary of `value_type`.
+    /// Corresponds to [arrow_schema::DataType::Dictionary].
+    Dictionary {
+        key_type: Box<DataType>,
+        value_type: Box<DataType>,
+    },
+    /// Exact 128-bit width decimal value with precision and scale.
+    /// Corresponds to [arrow_schema::DataType::Decimal128].
+    Decimal128 {
+        precision: u8,
+        scale: i8,
+    },
+    /// Exact 256-bit width decimal value with precision and scale
+    /// Corresponds to [arrow_schema::DataType::Decimal256].
+    Decimal256 {
+        precision: u8,
+        scale: i8,
+    },
+    /// A Map is a logical-nested type that is represented as
+    /// `List<entries: Struct<key: K, value: V>>`
+    /// Corresponds to [arrow_schema::DataType::Map].
     Map {
         key_type: Box<DataType>,
         value_type: Box<DataType>,
-        value_contains_null: bool,
+        value_type_nullable: bool,
+        keys_are_sorted: bool,
     },
+    ///
+    /// Everything below this line is not part of the Arrow specification.
+    ///
     UserDefined {
         jvm_class: Option<String>,
         python_class: Option<String>,
         serialized_python_class: Option<String>,
         sql_type: Box<DataType>,
     },
+    /// Resolves to either [DataType::Utf8] or [DataType::LargeUtf8],
+    /// based on `config.arrow_use_large_var_types`.
+    /// Optional length parameter is currently unused but retained for potential future use.
+    /// Optional `ConfiguredUtf8Type` specifies if the type is a `varchar` or `char`.
+    // TODO: The [`sail-spark-connect`] crate uses `TryFrom` for mapping Spark types to Sail types,
+    //  whereas [`PlanResolver`] resolves `DataType` without `TryFrom`.
+    //  This makes [`PlanResolver::resolve_data_type`] more suitable here,
+    //  as it has access to `config` and can determine the type accordingly.
+    //  Refactor data type resolution in [`sail-spark-connect`] to avoid using `TryFrom`,
+    //  which would allow for removal of this variant.
```

Comment:
I realized that we probably have to keep this since the SQL transformer does not have access to `PlanConfig`.

Maybe we could have a consolidated `String` type and use an enum (similar to `TimeZoneInfo`) to control how it maps to the Arrow type. This would be future work though.